Data race in test logger

```
==================
WARNING: DATA RACE
Read at 0x00c000ad6a03 by goroutine 7550:
  testing.(*common).logDepth()
      /usr/local/go/src/testing/testing.go:1024 +0x504
  testing.(*common).log()
      /usr/local/go/src/testing/testing.go:1011 +0x7d
  testing.(*common).Log()
      /usr/local/go/src/testing/testing.go:1052 +0x55
  github.com/prometheus/prometheus/util/testutil.logger.Log()
      /__w/mimir/mimir/vendor/github.com/prometheus/prometheus/util/testutil/logging.go:33 +0x46
  github.com/go-kit/log.(*context).Log()
      /__w/mimir/mimir/vendor/github.com/go-kit/log/log.go:168 +0x4ba
  github.com/grafana/mimir/pkg/storage/ingest.(*KafkaLogger).Log()
      /__w/mimir/mimir/pkg/storage/ingest/logger.go:36 +0x2ef
  github.com/twmb/franz-go/pkg/kgo.(*wrappedLogger).Log()
      /__w/mimir/mimir/vendor/github.com/twmb/franz-go/pkg/kgo/logger.go:123 +0xc6
  github.com/twmb/franz-go/pkg/kgo.(*broker).connect()
      /__w/mimir/mimir/vendor/github.com/twmb/franz-go/pkg/kgo/broker.go:640 +0xd70
  github.com/twmb/franz-go/pkg/kgo.(*broker).loadConnection()
      /__w/mimir/mimir/vendor/github.com/twmb/franz-go/pkg/kgo/broker.go:522 +0x189
  github.com/twmb/franz-go/pkg/kgo.(*broker).handleReq()
      /__w/mimir/mimir/vendor/github.com/twmb/franz-go/pkg/kgo/broker.go:302 +0x13c
  github.com/twmb/franz-go/pkg/kgo.(*broker).handleReqs()
      /__w/mimir/mimir/vendor/github.com/twmb/franz-go/pkg/kgo/broker.go:286 +0xda
  github.com/twmb/franz-go/pkg/kgo.(*broker).do.gowrap1()
      /__w/mimir/mimir/vendor/github.com/twmb/franz-go/pkg/kgo/broker.go:260 +0xab

Previous write at 0x00c000ad6a03 by goroutine 4269:
  testing.tRunner.func1()
      /usr/local/go/src/testing/testing.go:1677 +0x8fa
  runtime.deferreturn()
      /usr/local/go/src/runtime/panic.go:605 +0x5d
  testing.(*T).Run.gowrap1()
      /usr/local/go/src/testing/testing.go:1743 +0x44

Goroutine 7550 (running) created at:
  github.com/twmb/franz-go/pkg/kgo.(*broker).do()
      /__w/mimir/mimir/vendor/github.com/twmb/franz-go/pkg/kgo/broker.go:260 +0x2c4
  github.com/twmb/franz-go/pkg/kgo.(*broker).waitResp()
      /__w/mimir/mimir/vendor/github.com/twmb/franz-go/pkg/kgo/broker.go:275 +0x1ea
  github.com/twmb/franz-go/pkg/kgo.(*Client).handleShardedReq.func2.1()
      /__w/mimir/mimir/vendor/github.com/twmb/franz-go/pkg/kgo/client.go:2252 +0x314

Goroutine 4269 (finished) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1743 +0x825
  github.com/grafana/mimir/pkg/storage/ingest.TestPartitionReader_WaitReadConsistencyUntilLastProducedOffset_And_WaitReadConsistencyUntilOffset()
      /__w/mimir/mimir/pkg/storage/ingest/reader_test.go:439 +0x3ea
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1690 +0x226
  testing.(*T).Run.gowrap1()
      /usr/local/go/src/testing/testing.go:1743 +0x44
==================
--- FAIL: TestPartitionReader_WaitReadConsistencyUntilLastProducedOffset_And_WaitReadConsistencyUntilOffset (0.00s)
    logging.go:33: level warn partition 0 component kafka_client msg unable to open connection to broker addr 127.0.0.1:41709 broker 0 err dial tcp 127.0.0.1:41709: connect: connection refused
    --- FAIL: TestPartitionReader_WaitReadConsistencyUntilLastProducedOffset_And_WaitReadConsistencyUntilOffset/should_return_if_no_records_have_been_produced_yet (0.00s)
        --- FAIL: TestPartitionReader_WaitReadConsistencyUntilLastProducedOffset_And_WaitReadConsistencyUntilOffset/should_return_if_no_records_have_been_produced_yet/with_offset_true (0.03s)
            logging.go:33: level info partition 0 msg starting consumption from partition start because no offset has been found start_offset -2 consumer_group test-group
            logging.go:33: level info partition 0 component kafka_client msg immediate metadata update triggered why querying metadata for consumer initialization
            logging.go:33: level info partition 0 target_lag 2s max_lag 15s msg partition reader is starting to consume partition until target and max consumer lag is honored
            logging.go:33: level info partition 0 component kafka_client msg assigning partitions why new assignments from direct consumer how assigning everything new, keeping current assignment input test[0{-2 e-1 ce0}]
            logging.go:33: level info partition 0 target_lag 2s max_lag 15s msg partition reader found no records to consume because partition is empty partition_start_offset 0 last_produced_offset -1
            logging.go:33: level info partition 0 target_lag 2s max_lag 15s msg partition reader consumed partition and current lag is lower than configured target consumer lag last_consumed_offset -1 current_lag 0s
            logging.go:33: level debug partition 0 caller reader.go:622 msg waiting for read consistency
            logging.go:33: level debug partition 0 caller reader.go:643 msg catching up with offset offset -1
            logging.go:33: level info partition 0 msg stopping partition reader
            testing.go:1399: race detected during execution of test
        --- FAIL: TestPartitionReader_WaitReadConsistencyUntilLastProducedOffset_And_WaitReadConsistencyUntilOffset/should_return_if_no_records_have_been_produced_yet/with_offset_false (0.12s)
            logging.go:33: level info partition 0 msg starting consumption from partition start because no offset has been found start_offset -2 consumer_group test-group
            logging.go:33: level info partition 0 component kafka_client msg immediate metadata update triggered why querying metadata for consumer initialization
            logging.go:33: level info partition 0 target_lag 2s max_lag 15s msg partition reader is starting to consume partition until target and max consumer lag is honored
            logging.go:33: level info partition 0 component kafka_client msg assigning partitions why new assignments from direct consumer how assigning everything new, keeping current assignment input test[0{-2 e-1 ce0}]
            logging.go:33: level info partition 0 target_lag 2s max_lag 15s msg partition reader found no records to consume because partition is empty partition_start_offset 0 last_produced_offset -1
            logging.go:33: level info partition 0 target_lag 2s max_lag 15s msg partition reader consumed partition and current lag is lower than configured target consumer lag last_consumed_offset -1 current_lag 0s
            logging.go:33: level debug partition 0 caller reader.go:622 msg waiting for read consistency
            logging.go:33: level debug partition 0 caller reader.go:643 msg catching up with offset offset -1
            logging.go:33: level info partition 0 msg stopping partition reader
            testing.go:1399: race detected during execution of test
    --- FAIL: TestPartitionReader_WaitReadConsistencyUntilLastProducedOffset_And_WaitReadConsistencyUntilOffset/should_return_after_all_produced_records_have_been_consumed (0.00s)
        --- FAIL: TestPartitionReader_WaitReadConsistencyUntilLastProducedOffset_And_WaitReadConsistencyUntilOffset/should_return_after_all_produced_records_have_been_consumed/with_offset_true (1.04s)
            logging.go:33: level info partition 0 msg starting consumption from partition start because no offset has been found start_offset -2 consumer_group test-group
            logging.go:33: level info partition 0 component kafka_client msg immediate metadata update triggered why querying metadata for consumer initialization
            logging.go:33: level info partition 0 target_lag 2s max_lag 15s msg partition reader is starting to consume partition until target and max consumer lag is honored
            logging.go:33: level info partition 0 component kafka_client msg assigning partitions why new assignments from direct consumer how assigning everything new, keeping current assignment input test[0{-2 e-1 ce0}]
            logging.go:33: level info partition 0 target_lag 2s max_lag 15s msg partition reader found no records to consume because partition is empty partition_start_offset 0 last_produced_offset -1
            logging.go:33: level info partition 0 target_lag 2s max_lag 15s msg partition reader consumed partition and current lag is lower than configured target consumer lag last_consumed_offset -1 current_lag 0s
            reader_test.go:257: consumed record: record-1
            reader_test.go:268: produced 2 records (last record offset: 1)
            reader_test.go:272: started waiting for read consistency
            logging.go:33: level debug partition 0 caller reader.go:622 msg waiting for read consistency
            logging.go:33: level debug partition 0 caller reader.go:643 msg catching up with offset offset 1
            logging.go:33: level debug partition 0 msg last commit offset successfully committed to Kafka offset 0
            reader_test.go:257: consumed record: record-2
            reader_test.go:281: finished waiting for read consistency
            logging.go:33: level info partition 0 msg stopping partition reader
            logging.go:33: level debug partition 0 msg last commit offset successfully committed to Kafka offset 1
            testing.go:1399: race detected during execution of test
    --- FAIL: TestPartitionReader_WaitReadConsistencyUntilLastProducedOffset_And_WaitReadConsistencyUntilOffset/should_block_until_the_request_context_deadline_is_exceeded_if_produced_records_are_not_consumed (0.00s)
        --- FAIL: TestPartitionReader_WaitReadConsistencyUntilLastProducedOffset_And_WaitReadConsistencyUntilOffset/should_block_until_the_request_context_deadline_is_exceeded_if_produced_records_are_not_consumed/with_offset_false (1.12s)
            logging.go:33: level info partition 0 msg starting consumption from partition start because no offset has been found start_offset -2 consumer_group test-group
            logging.go:33: level info partition 0 component kafka_client msg immediate metadata update triggered why querying metadata for consumer initialization
            logging.go:33: level info partition 0 target_lag 2s max_lag 15s msg partition reader is starting to consume partition until target and max consumer lag is honored
            logging.go:33: level info partition 0 component kafka_client msg assigning partitions why new assignments from direct consumer how assigning everything new, keeping current assignment input test[0{-2 e-1 ce0}]
            logging.go:33: level info partition 0 target_lag 2s max_lag 15s msg partition reader found no records to consume because partition is empty partition_start_offset 0 last_produced_offset -1
            logging.go:33: level info partition 0 target_lag 2s max_lag 15s msg partition reader consumed partition and current lag is lower than configured target consumer lag last_consumed_offset -1 current_lag 0s
            reader_test.go:314: produced 1 record (last record offset: 0)
            logging.go:33: level debug partition 0 caller reader.go:622 msg waiting for read consistency
            logging.go:33: level debug partition 0 caller reader.go:643 msg catching up with offset offset 0
            logging.go:33: level debug partition 0 caller reader.go:622 msg waiting for read consistency
            logging.go:33: level debug partition 0 caller reader.go:643 msg catching up with offset offset 0
            logging.go:33: level info partition 0 msg stopping partition reader
            logging.go:33: level debug partition 0 msg last commit offset successfully committed to Kafka offset 0
            testing.go:1399: race detected during execution of test
FAIL
```

--- a/pkg/storage/ingest/reader_test.go
+++ b/pkg/storage/ingest/reader_test.go
@@ -27,6 +27,7 @@ import (
 	"github.com/twmb/franz-go/pkg/kmsg"
 	"go.uber.org/atomic"
 
+	mimirtest "github.com/grafana/mimir/pkg/util/test"
 	"github.com/grafana/mimir/pkg/util/testkafka"
 )
 
@@ -1713,7 +1714,7 @@ func withLogger(logger log.Logger) func(cfg *readerTestCfg) {
 func defaultReaderTestConfig(t *testing.T, addr string, topicName string, partitionID int32, consumer recordConsumer) *readerTestCfg {
 	return &readerTestCfg{
 		registry:    prometheus.NewPedanticRegistry(),
-		logger:      testutil.NewLogger(t),
+		logger:      mimirtest.NewTestingLogger(t),
 		kafka:       createTestKafkaConfig(addr, topicName),
 		partitionID: partitionID,
 		consumer:    consumer,

--- a/pkg/util/test/logger.go
+++ b/pkg/util/test/logger.go
@@ -3,6 +3,7 @@
 package test
 
 import (
+	"sync"
 	"testing"
 	"time"
 
@@ -10,7 +11,8 @@ import (
 )
 
 type testingLogger struct {
-	t testing.TB
+	t   testing.TB
+	mtx sync.Mutex
 }
 
 func NewTestingLogger(t testing.TB) log.Logger {
@@ -22,6 +24,10 @@ func NewTestingLogger(t testing.TB) log.Logger {
 func (l *testingLogger) Log(keyvals ...interface{}) error {
 	// Prepend log with timestamp.
 	keyvals = append([]interface{}{time.Now().String()}, keyvals...)
+
+	l.mtx.Lock()
 	l.t.Log(keyvals...)
+	l.mtx.Unlock()
+
 	return nil
 }

